# Hands-On Exercise 3: Engineering & Scientific Computing - Trigonometric Series and Convergence

####  Problem Scenario
I am supporting an engineering/scientific computing team working with a mathematical series. The team wants to understand how a series approximation behaves as more terms are included in a calculation. This notebook uses NumPy to generate numerical sequences, evaluate trigonometric fractions, and observe the concept of convergence.

####  Objectives
In this notebook, I will use NumPy to:
1. Convert angles from degrees to radians using `np.radians()`.
2. Generate continuous discrete tracking sequences using `np.arange()`.
3. Leverage broadcasting to perform element-wise array division (\(\frac{\sin(\theta)}{k}\)).
4. Calculate and compare progressive partial sums to observe numerical convergence.

In [1]:
import numpy as np

# Define angle in degrees
theta = 30

# Convert angle to radians
theta_radians = np.radians(theta)

print(f"Angle in Degrees: {theta}")
print(f"Angle in Radians: {theta_radians:.6f}")


Angle in Degrees: 30
Angle in Radians: 0.523599


In [2]:
# Create sequence for k from 1 up to (but excluding) 100
k = np.arange(1, 100, 1)

print("Generated k array:\n", k)
print("\n--- Array Inspection ---")
print("First value:", k[0])
print("Last value:", k[-1])
print("Number of values generated:", len(k))


Generated k array:
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]

--- Array Inspection ---
First value: 1
Last value: 99
Number of values generated: 99


In [3]:
# Calculate sine of the radian angle
sin_component = np.sin(theta_radians)

print(f"sin({theta}°) component value: {sin_component}")


sin(30°) component value: 0.49999999999999994


In [4]:
# Create individual terms using NumPy broadcasting (scalar / vector)
series_terms = sin_component / k

# Calculate the sum of the generated terms
total_sum = np.sum(series_terms)

print("First 5 terms of the series:", series_terms[:5])
print(f"Sum of terms (for k=1 to 99): {total_sum:.6f}")


First 5 terms of the series: [0.5        0.25       0.16666667 0.125      0.1       ]
Sum of terms (for k=1 to 99): 2.588689


In [5]:
# Evaluate the series across different range thresholds
sum_99 = np.sum(sin_component / np.arange(1, 100, 1))
sum_999 = np.sum(sin_component / np.arange(1, 1000, 1))
sum_9999 = np.sum(sin_component / np.arange(1, 10000, 1))

print("Progressive range evaluations completed successfully.")


Progressive range evaluations completed successfully.


### Convergence Investigation Table

| Number of Terms | Series Result |
| :--- | :--- |
| **99** | 2.588689 |
| **999** | 3.742235 |
| **9,999** | 4.893753 |

#### Observation:
As the number of terms increases, the result continues to grow, but the *rate* at which it grows slows down dramatically. Each subsequent order of magnitude adds a smaller fractional contribution to the total running sum.


### What Convergence Means in This Experiment

* **Fewer Terms:** When using only 99 terms, we get a rough, early approximation. The baseline calculation leaves out thousands of tiny subsequent fractions.
* **Increased Terms:** Moving to 9,999 terms includes values where $k$ is very large. This makes individual elements ($\frac{\sin(\theta)}{k}$) incredibly close to zero.
* **Why the Approximation Changes:** Each added term expands our precision. Because the denominator ($k$) grows continually while the numerator ($\sin(30^\circ) = 0.5$) stays completely constant, the terms being added shrink closer and closer to zero.
* **Limiting Value Meaning:** In scientific computing, approaching a limiting value means the algorithm is stabilizing. Even though we are adding thousands of new inputs, they become too small to significantly shift our primary calculations. This allows engineers to safely stop calculating once a desired level of numerical precision is reached.
